### ISSUE TITLE
Speed up CI workflows

### ISSUE THREAD

**Issue by yaalsn** [apache/pulsar#14401]
**Is your enhancement request related to a problem? Please describe.**
When we open a new PR, the CI workflows will run about one to two hours. So I wrote a script to analyze the time every workflow spends. The key is CI workflow name and the value is time(seconds)

```json
{
    "CI - Unit - Brokers - Broker": 13255,
    "CI - Unit": 5160,
    "CI - Unit - Brokers - Broker Group 1": 3939,
    "CI - Unit - Brokers - Other": 3622,
    "CI - Integration - Pulsar-IO Sinks and Sources": 3620,
    "CI - Integration - Cli": 3620,
    "CI - CPP build on Windows": 3619,
    "CI - Integration - Process": 3617,
    "CI - Integration - Thread": 3617,
    "CI - Unit - Brokers - Broker Group 2": 3090,
    "CI - Integration - Function State": 3024,
    "CI - Integration - Sql": 2816,
    "CI - Python - Build 3.9 client": 2762,
    "CI - Integration - Messaging": 2651,
    "CI - Integration - Backwards Compatibility": 2577,
    "CI - Integration - Pulsar-IO Oracle Source": 2563,
    "CI - Unit - Brokers - Publish - Throttle": 2531,
    "CI - Shade - Test": 2500,
    "CI - Unit - Brokers - Transaction": 2492,
    "CI - Unit - Flaky": 2441,
    "CI - Integration - Function & IO": 2432,
    "CI - Unit - Brokers - Flaky": 2308,
    "CI - Unit - Broker Auth SASL": 2274,
    "CI - Integration - Standalone": 2235,
    "CI - Integration - Schema": 2169,
    "CI - Integration - Transaction": 2074,
    "CI - Integration - Tiered FileSystem": 2047,
    "CI - Docker Build": 2023,
    "CI - Integration - Tiered JCloud": 1995,
    "CI - Unit - Brokers - Broker Group": 1868,
    "CI - Unit - Brokers - Others": 1637,
    "CI - Build - MacOS": 1513,
    "CI - Unit - Brokers - Long - Time": 1243,
    "CI - CPP, Python Tests": 1117,
    "CI - Unit - Adaptors": 1084,
    "CI - Unit - Brokers - Client Impl": 993,
    "CI - Misc": 969,
    "CI - Unit - Brokers - Client Api": 949,
    "CI - Unit - Proxy": 799,
    "CI - Unit - Brokers - Default": 699,
    "CI - CPP build on CentOS 7": 608,
    "CI - Unit - Broker - JDK8": 455,
    "Pulsar Bot": 213,
    "Auto Labeling": 201,
    "CI - Go Functions style check": 157,
    "CI - Go Functions Tests": 157,
    "CI - Maven Dependency Cache Update": 155,
    "CI - Pulsar - Build - 2.6": 141,
    "CI - Cancel duplicate workflows": 46,
    "CI - Misc - OWASP Dependency Check": 32,
    "CI - Deployment - Helm": 31
}
```

---

As we can see that `CI - Integration xxx`, `CI - Unit xxx` and  `CPP build on Windows` cost most. Let's step into these worfklows.

- `CI - Integration xxx`

Like all the other `CI - Integration xxx`, `CI - Integration - Pulsar-IO Sinks and Sources`  can be divided into three parts:

<img width="1019" alt="image" src="https://user-images.githubusercontent.com/10069311/154969970-bac4ae3e-3f90-4e5c-8d9b-e9003f0d61ec.png">

- `CI - Unit xxx`

<img width="1027" alt="image" src="https://user-images.githubusercontent.com/10069311/154971162-9dc0879f-2c4f-4454-a5cb-a93b4e0cfc58.png">

<img width="1023" alt="image" src="https://user-images.githubusercontent.com/10069311/154971403-3507fa10-d189-45ed-ae28-8802ae0deea4.png">

- `CPP build on Windows` 

<img width="1015" alt="image" src="https://user-images.githubusercontent.com/10069311/154973610-c6a54f6e-51b2-4849-93d4-96ca5d6a549b.png">

---

- All these `Intergration` and `Unit` jobs can be abstracted into `Build` and `Test`. `Build` contains `mvn install without test` and `docker build`, and `Test` is `mvn test`. There are so many workflow jobs execute `Build` step and cost most.
- `vcpkg install`  dependencies cost most.

**Describe the solution you'd like**

- Share maven install result files between all `Intergration` and `Unit` jobs. So we need to make a workflow to pre cache the result files first and then trigger the jobs. But consider that `Github Action` cannot re-run the fail steps now, when a job runs with fail, we need to re-run the whole workflow. So we also need to do some tricks to make the workflow to be idempotent.
- Cache the vcpkg dependencies.
- Use docker base images when docker build.

**Describe alternatives you've considered**
Remove useless or repeat jobs.

**Additional context**

**Comment by lhotari** [comment]
@yaalsn There's a proposal and proof-of-concept that I performed about a year ago: https://docs.google.com/document/d/1FNEWD3COdnNGMiryO9qBUW_83qtzAhqjDI5wwmPD-YE/edit . I'll soon resume this work since now GitHub support re-runing failed jobs in a workflow. That was the showstopper in making changes to the workflow.

**Comment by yaalsn** [comment]
> @yaalsn There's a proposal and proof-of-concept that I performed about a year ago: https://docs.google.com/document/d/1FNEWD3COdnNGMiryO9qBUW_83qtzAhqjDI5wwmPD-YE/edit . I'll soon resume this work since now GitHub support re-runing failed jobs in a workflow. That was the showstopper in making changes to the workflow.

Hi @lhotari. Thanks for your work. I just looked at the proposal and It is perfect! And I have two questions: 

1. `The GitHub Actions Cache can be used as a general purpose file storage to share the build artifacts as files between the stages`

    Do you mean `Github Action Artifacts` ?

2. I see `http://localhost:12321` in the [script](https://github.com/lhotari/pulsar/blob/lh-refactor-pulsar-ci-with-retries/build/ci_build_functions.sh). Is it a builtin url in github action?

**Comment by lhotari** [comment]
>    Do you mean `Github Action Artifacts` ?

No. I mean the GitHub Actions Cache. The cache is 5GB scoped for each fork which makes it optimal for caching build results of intermediate build steps in PR builds.

> I see `http://localhost:12321` in the [script](https://github.com/lhotari/pulsar/blob/lh-refactor-pulsar-ci-with-retries/build/ci_build_functions.sh). Is it a builtin url in github action?

No. It's part of [HTTP Cache Proxy action](https://github.com/marketplace/actions/http-cache-proxy), https://github.com/cirruslabs/http-cache-action.

The action is run in the build job and that starts the proxy on port `12321`:
https://github.com/lhotari/pulsar/blob/178eebdd97c6b9fff6b460089bdc3d8d5e4337a3/.github/workflows/pulsar-ci.yaml#L96-L97

The proxy uses the GitHub Actions Cache backend for persistence. HTTP POST (or PUT) can be used to store files and a HTTP GET can be used to retrieve a file. This is what the https://github.com/cirruslabs/http-cache-action enables.

**Comment by lhotari** [comment]
@yaalsn  I'm sorry I edited your original question accidentially when replying.

**Comment by lhotari** [comment]
I have rebased the Pulsar CI refactoring and it's now in tested at https://github.com/lhotari/pulsar/pull/59 . There are multiple changes that need to be done before it can be opened as a PR to apache/pulsar .

**Comment by lhotari** [comment]
Here's an example of how the main workflow could look like: https://github.com/lhotari/pulsar/blob/lh-github-actions-workflow-refactoring-2022/.github/workflows/pulsar-ci.yaml

**Comment by yaalsn** [comment]
Now https://github.blog/changelog/2021-11-23-github-actions-cache-size-is-now-increased-to-10gb-per-repository/

Even so, the Github action cache still has a limit of repository and It seems that the cache may be expired when the workflow is running. So I think using cache is not reliable but we can have a try first.

**Comment by lhotari** [comment]
I fixed a problem in the http-cache-action,  https://github.com/cirruslabs/http-cache-action/pull/2 . I'm running the workflow with that fix included. There might be other issues, but progress is being made.

To understand what the http-cache-action does, there's more information in https://medium.com/cirruslabs/http-caching-for-github-actions-f06c2e4e22e8 .

> Unfortunately, the default [GitHub Action for Caching](https://github.com/actions/cache) supports only folder caching even though under the hood it does downloads/uploads from/to Azure via an HTTP API. Luckily the protocol is [somewhat open sourced](https://github.com/actions/toolkit/blob/main/packages/cache/src/internal/cacheHttpClient.ts) but not documented. It was still manageable to “reverse engineer and write a proxy server in Go that exposes an API for build systems and internally uses the Azure API for actual downloads/uploads.

The source code of the official GitHub Action cache contains the details of the protocol as referenced in the blog post.

**Comment by lhotari** [comment]
latest run: https://github.com/lhotari/pulsar/actions/runs/2016939054

**Comment by lhotari** [comment]
The workflow refactoring is now ready. I'll prepare a PR to apache/pulsar.

Here's a screenshot of the combined workflow:
![image](https://user-images.githubusercontent.com/66864/159629905-da466f2d-4769-4b3d-b0c1-c16ee3315e27.png)

**Comment by yaalsn** [comment]
@lhotari Thanks for your contribution! Please @ me in the PR and I'll take a look.

**Comment by lhotari** [comment]
@yaalsn Please review #14819
